File: solutions/python/2020/day06.py
import re
from common2020 import str_block, print_ans


def formatted_notes(my_notes) -> list:
    groups_RE = re.compile(r'(\w+\n?)+(\n|$)')
    organized_notes = groups_RE.finditer(my_notes)
    
    all_group_ans = [individual_ans(group) for group in organized_notes]

    return all_group_ans


def individual_ans(group) -> list:
    individual_RE =re.compile(r'(\w)+')
    individuals_ans = individual_RE.finditer(group.group())
    return [i.group() for i in individuals_ans]


def ans_1(group_ans):

    combined_ans = ""
    total = 0

    for i in group_ans:
        for j in i:
            combined_ans += j

        total += len(set(combined_ans))
        combined_ans = ""

    return total


def ans_2(group_ans):

    total = 0
    combined_ans = ""

    for people in group_ans:
        if len(people) == 1:
            total += len(people[0])
        else:
            for ind in people:
                combined_ans += ind

            unique_ans = set(combined_ans)

            for i in unique_ans:
                if combined_ans.count(i) == len(people):
                    total += 1
            
            combined_ans = ""

    return total


def main():
    my_notes = str_block(6)
    group_ans = formatted_notes(my_notes)
    pt1 = ans_1(group_ans)
    pt2 = ans_2(group_ans)
    print_ans(pt1, pt2)


main()

File: solutions/typescript/2020/day06.ts
import { input, print } from "../common.ts";

function puzzleInput(): string[][] {
  return input(2020, 6)
    .split("\n\n")
    .map((line) => line.split("\n"));
}

function groupConsensus(group: string[]): number {
  const ansTotals: Record<string, number> = {};
  group.map((ind) =>
    ind.split("")
      .map((ans: string) =>
        ansTotals[ans] ? ansTotals[ans] += 1 : ansTotals[ans] = 1
      )
  );

  return Object.keys(ansTotals).reduce(
    (total, key) => ansTotals[key] === group.length ? total += 1 : total,
    0,
  );
}

function pt1(puzzleInput: string[][]): number {
  return puzzleInput.reduce((count, groupAns) =>
    count += new Set(
      groupAns.map((indAns) => indAns.split("")).flat(),
    )
      .size, 0);
}

function pt2(puzzleInput: string[][]): number {
  return puzzleInput.reduce(
    (count, groupAns) => count += groupConsensus(groupAns),
    0,
  );
}

print(pt1(puzzleInput()), pt2(puzzleInput()));
